src/Acornima.Extras/ScopeInfo.cs

```diff
@@ -0,0 +1,53 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Acornima.Ast;
+
+namespace Acornima;
+
+public sealed class ScopeInfo
+{
+    private readonly int _id;
+    private readonly int _currentVarScopeId;
+    private readonly int _currentThisScopeId;
+
+    public ScopeInfo(Node node, in Scope scope, ReadOnlySpan<Scope> scopeStack)
+    {
+        AssociatedNode = node;
+        _id = scope.Id;
+        _currentVarScopeId = scope.CurrentVarScopeIndex == scopeStack.Length ? scope.Id : scopeStack[scope.CurrentVarScopeIndex].Id;
+        _currentThisScopeId = scope.CurrentThisScopeIndex == scopeStack.Length ? scope.Id : scopeStack[scope.CurrentThisScopeIndex].Id;
```

Comment:
I think there may be an issue with the calculations for `_currentVarScopeId` and `_currentThisScopeId`, unless I'm misunderstanding their intended behavior (apologies if that's the case).

For `_currentThisScopeId`, it should reference the scope that encompasses the this/super references. For example, consider the following class:
```js
class B {
    constructor(ab) {
        this.ab = ab;
    }
    test() {
        this.ab = 3;
    }
}
```


**Expected behavior:**
The `_currentThisScopeField` for both the constructor and test methods should be associated with the scope created by the class B.
Each method should then have a distinct `_currentVarScopeId` reflecting their separate functional scopes.
```
Scope Program (B) scope ids: (0,0)
 └─ClassDeclaration
    ├─Identifier
    └─ClassBody
       ├─MethodDefinition
       │  ├─Identifier
       │  └─Scope FunctionExpression (ab) scope ids: (1,2)
       │     ├─Identifier
       └─MethodDefinition
          ├─Identifier
          └─Scope FunctionExpression () scope ids: (1,3)
```


**Current behavior:**
ClassDeclaration does not create a new scope and thus, the functions in the class B don't share the same `_currentThisScopeId`
```
Scope Program (B) scope ids: (0,0)
 └─ClassDeclaration
    ├─Identifier
    └─ClassBody
       ├─MethodDefinition
       │  ├─Identifier
       │  └─Scope FunctionExpression (ab) scope ids: (1,1)
       │     ├─Identifier
       └─MethodDefinition
          ├─Identifier
          └─Scope FunctionExpression () scope ids: (2,2)
```

Reply:
> The _currentThisScopeField for both the constructor and test methods should be associated with the scope created by the class B.

Classes don't create variable scopes as you can't directly declare variables in the body of a class, just properties and methods.

> I think there may be an issue with the calculations for _currentVarScopeId and _currentThisScopeId, unless I'm misunderstanding their intended behavior (apologies if that's the case).

You probably know that in JS `var` (and non-lexical `function`) declarations are hoisted to the top of the containing function body (or the top-level scope if there's none). Acorn uses the term *var scope* for referring to these scopes. (For more on hoisting, see e.g. [MDN](https://developer.mozilla.org/en-US/docs/Glossary/Hoisting)).

On the contrary, *this scope* means the scope which introduced the current "this" context, i.e. the context that the `this` keyword refers to. In practice, *this scopes* are almost the same as *var scopes*: top level scope and function bodies create *this scopes* EXCEPT FOR arrow functions, which may create a *var scope* but not a *this scope*. For example, in a script like `function f() { (() => { var x = this.x; console.log(x) })() }` the `x` variable is scoped to the arrow function's body but the `this` keyword refers to the this context of the `f` function. (More on this e.g. on [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/this)).

As far as I can see, Acorn deals with these concepts correctly.

BTW, the parser uses *this scope* for [figuring out a bunch of rules](https://github.com/adams85/acornima/blob/v1.0.0/src/Acornima/Parser.State.cs#L268-L317) like where the `super` keyword is allowed, etc. In your opinion, would it be useful to expose such info as well?

Reply:
>  this scope means the scope which introduced the current "this" context, i.e. the context that the this keyword refers to. In practice, this scopes are almost the same as var scopes: top level scope and function bodies create this scopes EXCEPT FOR arrow functions, which may create a var scope but not a this scope.

This makes more sense, I'm studying the ECMA have yet to fully grasp some of the details around JS, thank you for clarifying

Reply:
> BTW, the parser uses this scope for [figuring out a bunch of rules](https://github.com/adams85/acornima/blob/v1.0.0/src/Acornima/Parser.State.cs#L268-L317) like where the super keyword is allowed, etc. In your opinion, would it be useful to expose such info as well?

I can't think of any scenario where I would need them but, I only reached the surface of the current API, I will follow-up if I can think of cases where those could be useful